### PR TITLE
flash-attn integration

### PR DESCRIPTION
**Pull request by cyanguwa**
This PR is to provide support for flash attention in TE for self-attention calculations.

### CODE REVIEW

**Comment by ksivaman** [comment]
@cyanguwa Could you sign off the commits? [This guide](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst) explains how to do it.

File: setup.py [hunk 1]
```diff
@@ -313,5 +313,9 @@ def get_outputs(self):
     description="Transformer acceleration library",
     ext_modules=ext_modules,
     cmdclass={"build_ext": TEBuildExtension},
+    install_requires = [
+        "einops",
```
Comment (ptrendx) [inline setup.py:317]: I'm not 100% sure we need einops - in the end the 3 usages of rearrange we have are just glorified views...

File: transformer_engine/pytorch/transformer.py [hunk 2]
```diff
@@ -219,6 +228,243 @@ def forward(
...
+                number of key-value channels.
+    attention_dropout: float, default = 0.0
+                      dropout probability for the dropout op during multi-head attention.
+    layer_number: int, default = `None`
+                 layer number of the current `DotProductAttention` when multiple such modules
+                 are concatenated, for instance in consecutive transformer blocks.
+    apply_query_key_layer_scaling: bool, default = `True`
+                                  apply query-key layer scaling during BMM1
+                                  by a factor of `layer_number`
+    attention_softmax_in_fp32: bool, default = `True`
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:355]: Do we really need that option in the API?

File: transformer_engine/pytorch/transformer.py [hunk 3]
```diff
@@ -219,6 +228,243 @@ def forward(
...
+        tp_size: int = 1,
+        get_rng_state_tracker: Optional[Callable] = None,
+    ) -> None:
+        super().__init__()
+
+        self.use_flash_attention = int(os.getenv("NVTE_FLASH_ATTN", "1"))
+        if self.use_flash_attention:
+            attention = FlashAttention
+        else:
+            attention = UnfusedDotProductAttention
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:388]: I think this logic needs to be more sophisticated - FlashAttention supports only some input types/attention mask type etc. so we need to have a fallback

File: transformer_engine/pytorch/transformer.py (hunk 2 above)
Comment (ksivaman) [reply transformer_engine/pytorch/transformer.py:355]: It is a part of the `TransformerLayer` API, and as a result we pass it down here. If the question is more general about whether or not this is needed overall even in those, then maybe it's outside the scope of this PR?

File: transformer_engine/pytorch/transformer.py [hunk 4]
```diff
@@ -219,6 +228,243 @@ def forward(
...
+
+        self.attn_causal_mask = attn_mask_type == "causal"
+        self.apply_query_key_layer_scaling = apply_query_key_layer_scaling
+
+        if layer_number is None:
+            self.apply_query_key_layer_scaling = False
+        else:
+            self.layer_number = max(1, layer_number)
+
+        projection_size = kv_channels * num_attention_heads
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:271]: A lot of that logic is the same as the unfused attention - why not have that in the actual DotProductAttention class instead of repeated code here?

File: transformer_engine/pytorch/transformer.py (hunk 3 above)
Comment (ksivaman) [reply transformer_engine/pytorch/transformer.py:388]: Currently we throw an error when the user tries to use flash attn with the incorrect options. We could change that to a fallback. You're right in that we definitely need to handle the inputs in forward since flash attention does not support fp32.

File: transformer_engine/pytorch/transformer.py [hunk 5]
```diff
@@ -219,6 +228,243 @@ def forward(
...
+                        Boolean tensor used to mask out softmax input when not using flash-attn.
+        checkpoint_core_attention : bool, default = `True`
+                                   If true, forward activations for attention are recomputed
+                                   during the backward pass in order to save memory that would
+                                   otherwise be occupied to store the forward activations until
+                                   backprop.
+        """
+
+        if self.use_flash_attention:
+            assert attention_mask is None, "flash-attn does not require attention_mask."
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:452]: That should not be an assert - if the attention mask is here, you should just not use flash attention.

Or better yet - have mutually exclusive options for the attention_mask being boolean or a tensor of cumulative sequence lengths that flash attention needs (the unpadded case basically).

File: transformer_engine/pytorch/transformer.py [hunk 6]
```diff
@@ -343,8 +590,8 @@ def __init__(
                 **common_gemm_kwargs,
             )
 
-        # Core Self attention.
-        self.core_attention = CoreAttention(
+        # Attention.
+        self.dot_product_attention = DotProductAttention(
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:594]:
```suggestion
        self.core_attention = DotProductAttention(
```

From the point of view of the multiheadattention module that is its "core" attention mechanism :-).

File: transformer_engine/pytorch/transformer.py [hunk 7]
```diff
@@ -556,13 +772,23 @@ def forward(
         # core attention computation
         # ==================================
 
-        if checkpoint_core_attention:
-            context_layer = self._checkpointed_core_attention_forward(
-                query_layer, key_layer, value_layer, attention_mask
+        if self.use_flash_attention:
+            query_layer, key_layer, value_layer = [x.transpose(0,1).contiguous()
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:776]: This actually should be part of the FlashAttention forward function, that multiheadattention module should not really know whether the DotProductAttention uses FlashAttention or something else.

Comment (ptrendx) [reply transformer_engine/pytorch/transformer.py:776]: We need to have a contract with the user of `DotProductAttention` class what its inputs are.

File: transformer_engine/pytorch/transformer.py (hunk 2 above)
Comment (ptrendx) [reply transformer_engine/pytorch/transformer.py:355]: Fair enough. But yeah, we should ask ourselves whether it is a needed option there as well.

File: transformer_engine/pytorch/transformer.py [hunk 8]
```diff
@@ -211,14 +191,267 @@ def forward(
...
+            device=query_layer.device)
+
+        with self.attention_dropout_ctx():
+            output = flash_attn_unpadded_func(
+                query_layer, key_layer, value_layer, cu_seqlens, cu_seqlens, max_seqlen, max_seqlen,
+                self.attention_dropout if self.training else 0.0,
+                softmax_scale=1.0/self.norm_factor, causal=self.attn_causal_mask
+            )
+
+        # [b, sq, np, hn]
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:282]: Wrong comment or wrong code?

File: transformer_engine/pytorch/transformer.py [hunk 9]
```diff
@@ -211,14 +191,267 @@ def forward(
...
+        use_flash_attention = self.use_flash_attention
+        if (query_layer.dtype not in [torch.bfloat16, torch.float16]
+            or key_layer.dtype not in [torch.bfloat16, torch.float16]
+            or value_layer.dtype not in [torch.bfloat16, torch.float16]
+        ):
+            use_flash_attention = False
+
+        if attention_mask is not None:
+            use_flash_attention = False
+            warnings.warn(
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:435]: We should not warn if the `use_flash_attention` was already false before going into this `if`.

Comment (ptrendx) [reply transformer_engine/pytorch/transformer.py:435]: Also, I'm actually not sure we should warn at all - the user should not really care about it.

File: transformer_engine/pytorch/transformer.py [hunk 10]
```diff
@@ -211,14 +191,267 @@ def forward(
...
+            custom_forward,
+            False,
+            self.get_rng_state_tracker,
+            self.tp_group,
+            *forward_args,
+        )
+
+        return hidden_states
+
+    def forward(
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:398]: I would add some documentation here on the layout of q,k,v that we accept in this API.

File: transformer_engine/pytorch/transformer.py [hunk 11]
```diff
@@ -211,14 +191,267 @@ def forward(
...
+        ----------
+        query_layer : torch.Tensor
+                     Query tensor.
+        key_layer : torch.Tensor
+                   Key tensor.
+        value_layer : torch.Tensor
+                     Value tensor.
+        attention_mask : Optional[torch.Tensor], default = `None`
+                        Boolean tensor used to mask out softmax input when not using flash-attn.
+        checkpoint_core_attention : bool, default = `True`
```
Comment (ptrendx) [inline transformer_engine/pytorch/transformer.py:419]: Hmmm, maybe I'm mistaken, but shouldn't that be handled by this checkpoint context manager?

File: transformer_engine/pytorch/transformer.py (hunk 8 above)
Comment (ksivaman) [reply transformer_engine/pytorch/transformer.py:282]: The dimension comments are to indicate shape before not after

File: transformer_engine/pytorch/transformer.py (hunk 11 above)
Comment (ksivaman) [reply transformer_engine/pytorch/transformer.py:419]: The checkpoint context manager is called only when this flag is `True` and we go through the `_checkpointed_forward` path

Comment (ksivaman) [reply transformer_engine/pytorch/transformer.py:419]: But there is a mistake in the doc here since the default is `False`

File: transformer_engine/pytorch/transformer.py (hunk 8 above)
Comment (ptrendx) [reply transformer_engine/pytorch/transformer.py:281]: Well, then the code is wrong, since to get that you would need to do transpose, not view, right?

Comment (ksivaman) [reply transformer_engine/pytorch/transformer.py:281]: You are actually right, everything is wrong here; comment (input shape) and the view. I couldn't catch in my numerical tests since I'm running with mbs=1, sigh

**Review by ptrendx — APPROVED** [review]
LGTM, thanks!

**Comment by ksivaman** [comment]
ci error is unrelated ONNX failure due to tolerance

File: transformer_engine/pytorch/transformer.py [hunk 12]
```diff
@@ -211,14 +190,279 @@ def forward(
...
+            (query_layer.dtype in [torch.float16, torch.bfloat16])
+            and (key_layer.dtype in [torch.float16, torch.bfloat16])
+            and (value_layer.dtype in [torch.float16, torch.bfloat16])
+            ), 'FlashAttention currently only supports FP16 and BF16.'
+        assert (
+            query_layer.is_cuda and key_layer.is_cuda and value_layer.is_cuda
+            ), 'FlashAttention currently only supports CUDA tensors.'
+        assert (
+            attention_mask is None
+        ), 'FlashAttention currently does not support attention mask.'
```
Comment (cyanguwa) [inline transformer_engine/pytorch/transformer.py:253]: FA calculates the causal attention mask inside itself so maybe here we meant 'FlashAttention doesn't support external attention mask'?

File: qa/L0_unittest/test.sh [hunk 13]
```diff
@@ -7,4 +7,5 @@ set -e
 : ${TE_PATH:=/opt/transformerengine}
 
 pip install pytest==6.2.5 onnxruntime==1.13.1
-pytest -v -s $TE_PATH/tests/*.py
+pytest -v -s $TE_PATH/tests/test_transformerengine.py $TE_PATH/tests/test_fp8.py
+NVTE_FLASH_ATTN=0 pytest -v -s $TE_PATH/tests/test_onnx_export.py
```
Comment (cyanguwa) [inline qa/L0_unittest/test.sh:11]: We don't test NVTE_FLASH_ATTN=1 I guess for ONNX?

File: transformer_engine/pytorch/transformer.py [hunk 14]
```diff
@@ -211,14 +190,279 @@ def forward(
...
+        assert (
+            attention_softmax_in_fp32
+            ), 'FlashAttention currently only supports softmax compute in fp32.'
+
+        self.attn_causal_mask = attn_mask_type == "causal"
+        self.norm_factor = norm_factor
+        self.attention_dropout_ctx = attention_dropout_ctx
+        self.attention_dropout = attention_dropout
+        self.layer_number = layer_number
+        self.apply_query_key_layer_scaling = apply_query_key_layer_scaling
```
Comment (cyanguwa) [inline transformer_engine/pytorch/transformer.py:232]: Do we need L231-232 since we don't use layer_number and apply_query_key_layer_scaling in FA? I guess you're trying to keep the interface consistent between Unfused and Flash?

File: transformer_engine/pytorch/transformer.py [hunk 15]
```diff
@@ -211,14 +190,279 @@ def forward(
...
+        if (attention_mask is not None
+            or query_layer.dtype not in [torch.bfloat16, torch.float16]
+            or key_layer.dtype not in [torch.bfloat16, torch.float16]
+            or value_layer.dtype not in [torch.bfloat16, torch.float16]
+        ):
+            use_flash_attention = False
+
+        if use_flash_attention:
+            if checkpoint_core_attention:
+                return self._checkpointed_attention_forward(query_layer, key_layer, value_layer)
```
Comment (cyanguwa) [inline transformer_engine/pytorch/transformer.py:453]: Did we forget to pass in the attention_func here? Same for L457.

File: transformer_engine/pytorch/transformer.py (hunk 14 above)
Comment (ksivaman) [reply transformer_engine/pytorch/transformer.py:232]: Yes the second point is correct, just for consistency

File: transformer_engine/pytorch/transformer.py (hunk 15 above)
Comment (ksivaman) [reply transformer_engine/pytorch/transformer.py:453]: Yes there are some bugs in the selective activation checkpointing due to the MLM integration that doesn't actually run this path when I test for it, thus I wrongly assumed it works, fixing it now. Also will add standalone tests for this separately. Good catch :)

File: transformer_engine/pytorch/transformer.py [hunk 16]
```diff
@@ -211,14 +190,279 @@ def forward(
...
+        """Forward method with activation checkpointing."""
+
+        def custom_forward(*inputs):
+            return attention_func(*inputs)
+
+        hidden_states = checkpoint(
+            custom_forward,
+            False,
+            self.get_rng_state_tracker,
+            self.tp_group,
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply transformer_engine/pytorch/transformer.py:401]: Same as above
